### ISSUE TITLE
ALTER TABLE UPDATE does not work for DateField and DateTimeField

### ISSUE THREAD

**Issue by zahlio** [yandex/ClickHouse#3363]
I have the following table: 

```
create table grank.keyword
(
	id UInt64 default CAST(0, 'UInt64'),
	domain_id UInt64 default CAST(0, 'UInt64'),
	keyword String default '',
	date_added Date default CAST('1970-01-01', 'Date'),
	location String default '',
	countrylocale_id UInt16 default CAST(0, 'UInt16'),
	search_engine_id UInt8 default 0,
	search_type UInt8 default 0,
	preferred_landing_page_id Nullable(UInt64),
	tags Array(String),
	starred UInt8 default 0,
	ignore_local_results UInt8 default 0,
	ignore_video_results UInt8 default 0,
	ignore_featured_snippet UInt8 default 0,
	ignore_in_share_of_voice UInt8 default 0,
	deleted UInt8 default 0,
	version DateTime default CAST(0, 'DateTime')
)
engine = MergeTree() PARTITION BY floor(domain_id, -3) ORDER BY id SETTINGS index_granularity = 8192;
```

And run the following query:

```
ALTER TABLE keyword UPDATE
keyword='some keyword',
date_added=toDate('2018-10-12'),
location='', countrylocale_id=76,
search_engine_id=1,
search_type=1,
preferred_landing_page_id=NULL,
tags=['this is tag'],
starred=0,
ignore_local_results=0,
ignore_video_results=0,
ignore_featured_snippet=0,
ignore_in_share_of_voice=0,
deleted=0,
version=toDateTime(1539333037)
WHERE id=2555 AND domain_id=9;
```

And get the following error:
```
[2018-10-12 10:50:06] Code: 44, e.displayText() = DB::Exception: Illegal columns Const(UInt16) and UInt16 of second (then) and third (else) arguments of function if, e.what() = DB::Exception
```

**Comment by 4ertus2** [comment]
Hi,

It's known regression that appears a few days ago. You may try previous build or wait a fix.

**Comment by zahlio** [comment]
We tried in stable and on test, both had the issue, but awesome you guys are working on a fix.

**Comment by 4ertus2** [comment]
First affected revision 94c9a544 (10.10.18)
Fixed in master in 00026f1b8 (12.10.18)
